containers/podman-desktop
# feat: dynamic breadcrumbs

### What does this PR do?

See issue #3114 for background.

In order to avoid duplicate/different breadcrumb strings, this change uses the existing Route breadcrumb as the name of each page. This requires a one-time change to some of the existing breadcrumbs to match the corresponding action/what we should actually call the page.

We have three special kinds of Routes:
- the main nav lists
- the details pages
- tabs

The first two are the only ones you can go 'back' to, and can only be one (list > action) or two (list > details > action) levels deep. Tabs affect the current URL but don't affect what the 'current' page is. A new navLevel property is added to tag these special routes.

In order to keep things relatively simple we'll just track the first two kinds of routes as properties in a store and set a lastPage whenever we enter a route. Likewise, the current page changes for any route except tabs, but detail routes are special since they need to change the back-link. For instance, if you click on Image Details > Inspect, then Run Image, cancelling should return you to Inspect instead of the image summary.

Someone may ask: why are breadcrumbs put in a store vs a Route property like meta? Because a visual page like Details is rendered from several routes (/, /containers/:id/, /containers/:id/xyz) and getting the main page to have the correct name and URL for a tab when we haven't gotten to that route yet is more complicated than simply sharing a store.

- Detailed change notes -

Route/breadcrumb:
- Use navLevel as described above to keep track of current list/details pages and set last/current page in breadcrumb store.

App:
- Fix breadcrumb names and set basic navLevels.

Details pages:
- Use breadcrumb in DetailsPage, add navLevel to tabs and remove parent from *Details pages.

Form pages:
- Use breadcrumb in FormPage and remove parent from the two pages that use it.
- Added showBreadcrumb property so Help page can still turn it off.
- Set navLevel in RunImage so that the current page name doesn't pick up the tab name.

### Screenshot/screencast of this PR

<img width="277" alt="Screenshot 2023-07-05 at 9 16 08 AM" src="https://github.com/containers/podman-desktop/assets/19958075/98831863-ce07-4c0c-b774-6ceb4671982f">
<img width="277" alt="Screenshot 2023-07-05 at 9 16 31 AM" src="https://github.com/containers/podman-desktop/assets/19958075/b131fba8-a986-458f-8918-eb0641646284">
<img width="277" alt="Screenshot 2023-07-05 at 9 16 42 AM" src="https://github.com/containers/podman-desktop/assets/19958075/755c249a-bd38-4585-8595-86ba923613dc">

### What issues does this PR fix or reference?

Fixes issue #3114.

### How to test this PR?

Go to each details page to confirm breadcrumb looks correct. Go to containers list and select a pod to confirm the breadcrumb is now Containers > Pod Details.

Go to each of the form pages: Help (no breadcrumb), Run Image, Deploy Pod to Kubernetes to confirm the page name and breadcrumb at the top. Try these actions from both the list and details pages to confirm the breadcrumb is different and correct in both cases.

On the image details page select the History or Inspect tab before using Run Image, and confirm the breadcrumb or close goes back to the correct tab.
